### PR TITLE
[ENG-892] Source: Improve efficiency of realtime ingestion by not querying campaign_leads

### PR DESCRIPTION
**Pull request by dean-martin**
https://thedmsgrp.atlassian.net/browse/ENG-892

TODO: 
- Test to make sure this change doesn't break anything.

Testing process: 
- You just need an empty campaign fed by a Source,
- Post the new lead to that source and campaign using index_dev.php to get the logged queries.

Test without the change and then with it...

Not sure how to test not detaching if $this->realTime is true (Model\Api)....

We likely need to share screen to show you that process. let me/scott/nick know if/when you want to see that test process.

### CODE REVIEW

**Comment by dean-martin** [comment]
So far the change looks correct, using xdebug to test it. Still need to test without the change to make sure that's where it was querying.

**Comment by dean-martin** [comment]
This change is incorrect, before this change, the query was never hit the code that I changed. 
There's different places currently querying `campaign_leads` that I have to find.

**Comment by dean-martin** [comment]
The query seems to be in CampaignBundle/Entity/CampaignRepository::getPendingContactIds….

It doesn’t look like it’s possible to extend ContactLimiter to skip this query, the query is being built using some input from ContactLimiter, and only allowing to modify parts of the query… but I’ll keep looking

**Comment by dean-martin** [comment]
If we're not able to skip the query, we might be able to make the query significantly faster by setting $contactId on ContactLimiter. This'll set a WHERE lead_id = :contactId, making the query faster....

RealTimeContactLimiter will likely be deleted.

**Comment by dean-martin** [comment]
Update, going to do @heathdutton 's idea to create a FakeCampaignRepository that would replace CampaignRepository::getPendingContactIds to take the $contactIdList from ContactLimiter
OR a FakeKickoffContactFinder where getContacts would do the same.

**Comment by dean-martin** [comment]
Changes looks to be working.
Because this is a likely a risky change, I'd like to test it thoroughly before merging.

File: Entity/RealTimeCampaignRepository.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        if ($limiter->hasCampaignLimit() && $limiter->getCampaignLimitRemaining() < $limiter->getBatchLimit()) {
+            $pulled = [];
+            for ($i = $limiter->getCampaignLimitRemaining();
+                $i > ($limiter->getCampaignLimitRemaining() - $limiter->getBatchLimit()); --$i) {
+                if (isset($contacts[$i])) {
+                    $pulled[] = $contacts[$i];
+                }
+            }
+            $pulled = $contacts;
+        }
```
Comment (heathdutton) [inline Entity/RealTimeCampaignRepository.php]: Should be `$contacts = $pulled;` I think. There are more elegant ways of reducing the count of items in an array than a for loop.

```
$contacts = array_slice($contacts, 0, $limiter->getCampaignLimitRemaining());
```

**Comment by heathdutton** [comment]
Looks like it needs another PHPCS sweep, for raisins.

**Pull request merged**